ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java

```diff
@@ -0,0 +1,529 @@
+   private GpuMat sensorCroppedContactMapImage;
+
+   private boolean initialized = false;
+   private int mode = 1; // 0 -> Ouster, 1 -> Realsense
+   private float gridOffsetX;
+   private int centerIndex;
+   private int localCellsPerAxis;
+   private int globalCenterIndex;
+   private int cropCenterIndex;
+   private int globalCellsPerAxis;
+   private CameraIntrinsics cameraIntrinsics;
+
+   private TerrainMapData terrainMapData;
+   private static HeightMapParameters heightMapParameters = new HeightMapParameters("GPU");
+   private final SideDependentList<ReferenceFrame> footSoleFrames = new SideDependentList<>();
+
+   private Rect cropWindowRectangle;
+   private static final boolean computeSteppability = true;
+   private CUDAProgram heightMapCUDAProgram;
+   private CUstream_st stream;
+
+   private final RigidBodyTransform currentSensorToWorldTransform = new RigidBodyTransform();
+   private final RigidBodyTransform currentGroundToWorldTransform = new RigidBodyTransform();
+   private final Point3D sensorOrigin = new Point3D();
+   private final TerrainMapStatistics terrainMapStatistics = new TerrainMapStatistics();
+
+   private HeightMapAutoencoder denoiser;
+   private Mat denoisedHeightMapImage;
+   private Mat steppableRegionAssignmentMat;
+   private Mat steppableRegionRingMat;
+   public int sequenceNumber = 0;
+
+   private float[] worldToGroundTransformArray = new float[16];
+   private float[] groundToWorldTransformArray = new float[16];
+   private float[] groundToSensorTransformArray = new float[16];
+   private float[] sensorToGroundTransformArray = new float[16];
+
+   private FloatPointer groundToSensorTransformHostPointer;
+   private FloatPointer groundToSensorTransformDevicePointer;
+
+   private FloatPointer sensorToGroundTransformHostPointer;
+   private FloatPointer sensorToGroundTransformDevicePointer;
+
+   private FloatPointer worldToGroundTransformHostPointer;
+   private FloatPointer worldToGroundTransformDevicePointer;
+
+   private FloatPointer groundToWorldTransformHostPointer;
+   private FloatPointer groundToWorldTransformDevicePointer;
+
+   private FloatPointer parametersHostPointer;
+   private FloatPointer parametersDevicePointer;
+
+   private CUDAKernel updateKernel;
+   private CUDAKernel registerKernel;
+   private CUDAKernel croppingKernel;
+   private CUDAKernel preprocessKernel;
+
+   private dim3 blockSize;
+   private dim3 gridSizeKernel0;
+   private dim3 gridSizeKernel1;
+   private dim3 gridSizeKernel2;
+   private dim3 gridSizeKernel3;
+
+   float[] paramsArray;
+
+   private boolean processing = false;
+
+   public void setModeSpherical()
+   {
+      mode = 0;
+   }
+
+   public void setModePerspective()
+   {
+      mode = 1;
+   }
+
+   public RapidHeightMapExtractorCUDA(ReferenceFrame leftFootSoleFrame, ReferenceFrame rightFootSoleFrame)
+   {
+      footSoleFrames.put(RobotSide.LEFT, leftFootSoleFrame);
+      footSoleFrames.put(RobotSide.RIGHT, rightFootSoleFrame);
```

Comment:
Is it possible not to use RobotSide or reference any part of a robot in this? That way the height map extractor is very generic and can be reused without having to be specific to a humanoid robot.

Reply:
Yeah, it'd be nice if we could run the height map without a robot, especially for testing. Definitely a change that should be made in the future, but I don't think it's necessary for this PR.